[WebGPU] Loops are susceptible to UB optimizations

mwyrzykowski opened the pull request:
#### f7fc51937f2ee3eb0258f99408985316f8b3e4d5
<pre>
[WebGPU] Loops are susceptible to UB optimizations
<a href="https://bugs.webkit.org/show_bug.cgi?id=283595">https://bugs.webkit.org/show_bug.cgi?id=283595</a>
<a href="https://rdar.apple.com/140334491">rdar://140334491</a>

Reviewed by Dan Glastonbury.

Ensure loops always make forward progress otherwise UB can result
in the optimizer optimizing away necessary clamping functions.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
Emit volatile read to ensure forward progress.

* LayoutTests/fast/webgpu/regression/repro_283595-for-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-for-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-for.html: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop.html: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while.html: Added.
Add regression test.

Canonical link: <a href="https://commits.webkit.org/287260@main">https://commits.webkit.org/287260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a09858b670cfbbc270b335be37e136c0347d41a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57977 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32317 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83583 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30170 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6255 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61830 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19724 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82005 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71820 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42105 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49187 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26019 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28528 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26428 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84964 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6283 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4335 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70065 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6447 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69284 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17265 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13328 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12068 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6233 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6197 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9648 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7988 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on LayoutTests/fast/webgpu/nocrash/fuzz-283595-for-expected.txt:
```diff
@@ -0,0 +1 @@
+Test passes if it does not crash.
```
I don't think the expectations are correct?
The test cases should end with context loss if the browser works..

mwyrzykowski replied:
Why should the context get lost? Because context is lost when the device is destroyed?

kkinnunen-apple replied:
Because the test has a infinite loop and current implementation should preserve the infinite loop.
And the way to break the infinite loop is to reset the gpu.
And when the gpu is reset, the context is lost.

It definitively doesn't result in string "Test passes if it does not crash.", as the test does not contain such string.

mwyrzykowski replied:
I updated the expectations to be timeouts as expected

mwyrzykowski replied:
Going to switch these to vertex shaders as discussed offline and then they should result in context lost

mwyrzykowski replied:
I used a fragment shader instead which still triggers the GPU hang error and await the device lost promise in the expectations. Vertex shaders in WebGPU don't support writable storage buffers and while the repro could be modified to not use storage buffers, it wasn't immediately obvious how to test the out of bounds write did not occur.

kkinnunen-apple commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2422,13 +2422,14 @@ void FunctionDefinitionWriter::visit(AST::ForStatement& statement)
 void FunctionDefinitionWriter::visit(AST::LoopStatement& statement)
 {
-    m_stringBuilder.append("while (true) {\n"_s);
+    m_stringBuilder.append("while (true) { volatile bool __wgslEnsureForwardProgress; (void)__wgslEnsureForwardProgress; \n"_s);
```
The read of uninitialized variable is UB, so you're adding more UB here.
I'm not sure it makes sense to instantiate the variable inside the loop. When you add initializer, it will add a volatile write in the loop bodies. 

It's better to add the variable before the first loop, and make all nested loops use the same variable.

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] Loops are susceptible to UB optimizations
https://bugs.webkit.org/show_bug.cgi?id=283595
rdar://140334491

Reviewed by Dan Glastonbury.

Ensure loops always make forward progress otherwise UB can result
in the optimizer optimizing away necessary clamping functions.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
Emit volatile read to ensure forward progress.

* LayoutTests/fast/webgpu/regression/repro_283595-for-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-for-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-for.html: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-loop.html: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while-expected.png: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_283595-while.html: Added.
Add regression test.

Canonical link: https://commits.webkit.org/287260@main
